[Target][Lowering] Update Op Intrinsic Lowering Mechanism And Intrinsic Lowering Pass

zxybazh opened the pull request:
This PR updated the intrinsic lowering pass to support the new op registry and avoid overloading the global tvm registry. Meanwhile, it kept the fallback mechanism to find the most suitable lower intrinsic function, e.g., `llvm.FLowerIntrinsic` vs. `default.FLowerIntrinsic`. All previous op registration are ported to new functions, and some missing ops would be added in separate PR.

zxybazh commented:
Python side op lowering function has been changed from `register_func` to `register_op`.

junrushao1994 commented:
Thanks for the contribution! A few high-level comments:

1. Do not add any C API - we can use packed func to expose the interface. Op is a compile-time concept so please move the registration logic out of the runtime :-)
2. Please change `register_op` to some name like `register_intrin_lowering`, and consider accept a target kind like `default`/`llvm`/`cuda`

zxybazh commented:
Thanks for the advice! I have changed the name of function in python-side to `register_op_intrin_lowering` and removed change to C API. Instead, I utilized the global function registration and packedfunc to register op intrinsic lowering function.

zxybazh commented:
@junrushao1994 Please review, thanks!

junrushao1994 commented on include/tvm/tir/op_attr_types.h:
```diff
@@ -28,11 +28,13 @@
 #ifndef TVM_TIR_OP_ATTR_TYPES_H_
 #define TVM_TIR_OP_ATTR_TYPES_H_
 
+#include <tvm/ir/expr.h>
```
No need for this

junrushao1994 commented on include/tvm/tir/op_attr_types.h:
```diff
@@ -28,11 +28,13 @@
 namespace tvm {
 namespace tir {
-
+using namespace runtime;
```
No need for this

junrushao1994 commented on include/tvm/tir/op_attr_types.h:
```diff
@@ -43,6 +45,11 @@ using TGlobalSymbol = String;
+/*!
+ * \brief The intrinsic lowering function for given OP.
+ */
+using FLowerIntrinsic = PackedFunc;
```
Using `runtime::TypedPackedFunc` when possible, which is a simple wrapper of `PackedFunc`, but with the clear signature like `std::function`, so it could help readability :-)

junrushao1994 reviewed (commented):
Thanks for the contribution!

junrushao1994 commented on python/tvm/ir/op.py:
```diff
@@ -115,3 +127,50 @@ def _register(v):
+    fregister : function
+        Register op lowering function if f is not specified.
+    """
+    if not isinstance(op_name, str):
+        raise ValueError("expect string op name")
+
+    def _register(myf):
+        """internal intrinsic lowering registration function"""
+        assert isinstance(target, str)
+        if not isinstance(myf, PackedFuncBase):
+            myf = convert_to_tvm_func(myf)
+        _ffi_api.RegisterOpLowerIntrinsic(op_name, myf.handle, target, level, override)
+        return myf
```
I assume the logic could be a bit simpler? Perhaps you might want to refer the some previous implementations, like "def register_compute" :-)

zxybazh replied:
Sure it could be simplified but I'll still keep the `_register` function for consistensy in function implementation compared to `register_op_attr`.

junrushao1994 commented on src/ir/op.cc:
```diff
@@ -122,6 +124,19 @@ TVM_REGISTER_GLOBAL("ir.RegisterOpAttr")
     });
 
+TVM_REGISTER_GLOBAL("ir.RegisterOpLowerIntrinsic")
+    .set_body_typed([](String name, TVMFunctionHandle f, String target = "default", int plevel = 10,
+                       int can_override = 0) {
```
Forget about TVMFunctionHandle, which is only used in the C API but nowhere else :-)

Let's embrace the PackedFunc magic! It is a nicer wrapper of a function from either C++ or python (or potentially other languages).

BTW, don't need to set default arguments of a lambda because they are not used any way.

```suggestion
    .set_body_typed([](String name, PackedFunc f, String target, int plevel, bool override_) {
```

junrushao1994 commented on src/ir/op.cc:
```diff
@@ -122,6 +124,19 @@ TVM_REGISTER_GLOBAL("ir.RegisterOpAttr")
       }
     });
 
+TVM_REGISTER_GLOBAL("ir.RegisterOpLowerIntrinsic")
```
I assume this should be put on `src/tir/op.cc`?

zxybazh replied:
Yeah I have the same concern but I checked both files and found most similar functions like `RegisterOpAttr` function in the same file. Therefore, I think it better to keep the function here or we would need to migrate all related functions.

junrushao1994 commented on python/tvm/ir/op.py:
```diff
@@ -115,3 +127,50 @@ def _register(v):
     return _register(value) if value is not None else _register
+
+
+def register_op_intrin_lowering(
+    op_name,
+    f=None,
+    target="default",
+    level=10,
+    override=False,
+):
```
Let's avoid providing too many defaults :-)


```suggestion
def register_op_intrin_lowering(
    op_name,
    f,
    target,
    level=10,
    override=False,
):
```

junrushao1994 reviewed (changes requested):
Please find some further comments :-)

junrushao1994 commented on src/target/intrin_rule.cc:
```diff
@@ -24,179 +24,234 @@
+using tir::FLowerIntrinsic;
 
-TVM_REGISTER_GLOBAL("tvm.intrin.rule.default.exp").set_body(DispatchPureExtern<FloatSuffix>);
+TVM_REGISTER_OP("tir.exp").set_attr<FLowerIntrinsic>("default.FLowerIntrinsic",
```
shall we remove this file to, like, `lower_intrinsic_default.cc`? CC: @tqchen

zxybazh replied:
I think the naming is just fine because the `lower_intrinsic.cc`file kind of implies it's default rules.

tqchen commented:
note, please do not checkin vta-hw submodule change(do a git submodule update)

zxybazh commented:
> note, please do not checkin vta-hw submodule change(do a git submodule update)

Noted with thanks!

junrushao1994 commented on python/tvm/ir/op.py:
```diff
@@ -17,9 +17,21 @@
 from .expr import RelayExpr
 from . import _ffi_api
 
+try:
+    # pylint: disable=wrong-import-position,unused-import
+    if _FFI_MODE == "ctypes":
+        raise ImportError()
+    from tvm._ffi._cy3.core import convert_to_tvm_func, _get_global_func, PackedFuncBase
+except (RuntimeError, ImportError) as error:
+    # pylint: disable=wrong-import-position,unused-import
+    if _FFI_MODE == "cython":
+        raise error
+    from tvm._ffi._ctypes.packed_func import convert_to_tvm_func, _get_global_func, PackedFuncBase
```
😆 i suppose these lines become useless given you have addressed the comments i left before?

zxybazh replied:
Removed, thanks!

junrushao1994 commented on src/ir/op.cc:
```diff
@@ -122,6 +124,18 @@ TVM_REGISTER_GLOBAL("ir.RegisterOpAttr")
+      if (Op::HasAttrMap(target + ".FLowerIntrinsic") &&
+          OpRegistry::Global()->Get(name) != nullptr &&
+          Op::GetAttrMap<FLowerIntrinsic>(target + ".FLowerIntrinsic").count(Op::Get(name))) {
+        ICHECK(can_override) << "Op " << name << "'s intrinsic lowering function " << target
```
Use `CHECK` here because it is a user-facing error message.

```suggestion
        CHECK(can_override) << "Op " << name << "'s intrinsic lowering function " << target
```

junrushao1994 commented on src/ir/op.cc:
```diff
@@ -122,6 +124,18 @@ TVM_REGISTER_GLOBAL("ir.RegisterOpAttr")
+        ICHECK(can_override) << "Op " << name << "'s intrinsic lowering function " << target
+                             << ".FlowerIntrinsic is already registered";
+      }
+      tvm::OpRegEntry::RegisterOrGet(name).set_name().set_attr<FLowerIntrinsic>(
```
No need to set_name in this case

junrushao1994 commented on src/target/llvm/intrin_rule_llvm.cc:
```diff
@@ -25,155 +25,175 @@
-      using tir::make_zero;
+TVM_REGISTER_OP("tir.exp10")
+    .set_attr<FLowerIntrinsic>("llvm.FLowerIntrinsic",
+                               PackedFunc([](const TVMArgs& targs, TVMRetValue* rv) {
```
IIRC without this `PackedFunc` conversion the code still compiles (because we do implicit conversion), so in this particular case (registering with a lambda), we can remove this explicit conversion

zxybazh replied:
Unfortunately if I remove the type conversion I get the following error.
`tvm/src/target/llvm/intrin_rule_llvm.cc:196:7: note:   cannot convert ‘<lambda closure object>tvm::codegen::llvm::<lambda(const tvm::runtime::TVMArgs&, tvm::runtime::TVMRetValue*)>{}’ (type ‘tvm::codegen::llvm::<lambda(const tvm::runtime::TVMArgs&, tvm::runtime::TVMRetValue*)>’) to type ‘const tvm::runtime::TypedPackedFunc<tvm::PrimExpr(tvm::PrimExpr)>&’
     }));`

junrushao1994 replied:
oh i see...let's keep it then :-)

junrushao1994 reviewed (changes requested):
The PR overall is in a pretty good state. Please find my extra nitpicks below. Also, would you like to proofread that tutorial to make sure the instructions are up-to-date? Thanks a lot!

junrushao1994 commented on python/tvm/ir/op.py:
```diff
@@ -115,3 +127,43 @@ def _register(v):
     return _register(value) if value is not None else _register
+
+
+def register_op_intrin_lowering(
+    op_name,
+    target,
+    f=None,
+    level=10,
+    override=False,
+):
```
oh BTW, the order of the arguments matter, and we need to double check if the following syntax works

```
@tvm.ir.op.register_op_intrin_lowering("tir.exp", target="cuda", override=True)
def my_exp(expr) -> expr:
    pass
```

zxybazh replied:
Thanks for point that out. I thought about this issue again and I think here `f` is used overwhelmingly more than `level` and `override`. In most cases, it's just `op_name`, `target` and `f`. Therefore, I insist that we keep the current order for most intuitive function usage.

junrushao1994 replied:
An alternative approach is to force kwargs, as demonstrated below, so that users are required to explicitly mark which argument is override and which is level

```python
def (a, b, *, c, d)
```

zxybazh replied:
In that case, it would be awkward to use the function as decorators as we would need to mark all the argument. What do you think?

junrushao1994 replied:
This is an interesting topic to discuss :-)

In our particular case, we could still provide default values for some arguments, like:

```python
>>> def f(a, *, b, c=1):
...   print(f"a = {a}, b = {b}, c = {c}")
...
>>> f(1, 2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: f() takes 1 positional argument but 2 were given
>>> f(1, b=2)
a = 1, b = 2, c = 1
```

zxybazh replied:
Ah I see. With default values things could be a lot easier. I would prefer to use `def register_intrin_lowering(op_name, target, *, f=None, level=10, override=False,)`

junrushao1994 commented on src/tir/transforms/lower_intrin.cc:
```diff
@@ -42,28 +42,36 @@ class IntrinInjecter : public tvm::arith::IRMutatorWithAnalyzer {
-      std::string name = ptr_op->name;
-      PrimExpr r = ApplyPattern(name, GetRef<PrimExpr>(op));
-      if (r.defined()) return r;
+      for (size_t i = 0; i < patterns_.size(); ++i)
```
```suggestion
      for (const std::string& pattern : patterns_)
```

junrushao1994 commented on src/tir/transforms/lower_intrin.cc:
```diff
@@ -42,28 +42,36 @@ class IntrinInjecter : public tvm::arith::IRMutatorWithAnalyzer {
-      if (r.defined()) return r;
+      for (size_t i = 0; i < patterns_.size(); ++i)
+        if (Op::HasAttrMap(patterns_[i])) {
+          auto default_intrin = Op::GetAttrMap<FLowerIntrinsic>(patterns_[i]);
```
it is not necessarily default intrin, so let's find a more accurate name here? like, `f_lower_intrin_map`?

junrushao1994 commented on src/tir/transforms/lower_intrin.cc:
```diff
@@ -42,28 +42,36 @@ class IntrinInjecter : public tvm::arith::IRMutatorWithAnalyzer {
+        if (Op::HasAttrMap(patterns_[i])) {
+          auto default_intrin = Op::GetAttrMap<FLowerIntrinsic>(patterns_[i]);
+          FLowerIntrinsic f = default_intrin.get(GetRef<Op>(ptr_op), nullptr);
+          PrimExpr e = GetRef<PrimExpr>(op);
+          if (f != nullptr) {
+            PrimExpr r = f(e);
```
```suggestion
          if (f != nullptr) {
            PrimExpr e = GetRef<PrimExpr>(op);
            PrimExpr r = f(e);
```

junrushao1994 commented on src/tir/transforms/lower_intrin.cc:
```diff
@@ -42,28 +42,36 @@ class IntrinInjecter : public tvm::arith::IRMutatorWithAnalyzer {
+            ICHECK(r.defined()) << "intrinsic rule must always return valid Expr";
+            if (!r.same_as(e)) {
+              r = this->VisitExpr(r);
+              if (r.defined()) return r;
```
```suggestion
              if (r.defined()) {
                return r;
              }
```

junrushao1994 reviewed (approved):
I don't have other comments :-) Thanks for the contribution!

junrushao1994 commented:
CC: @comaniac @yzhliu @icemelon9

junrushao1994 commented:
@tqchen would you like to take a second look? If there is other comments, I will get it merged the other day

tqchen commented on include/tvm/ir/op.h:
```diff
@@ -270,14 +270,17 @@ class OpRegEntry {
   template <typename ValueType>
   inline OpRegEntry& set_attr(const std::string& attr_name,  // NOLINT(*)
-                              const ValueType& value, int plevel = 10);
+                              const ValueType& value, int plevel = 10, int can_override = 0);
```
Wju do we need to introduce can_override when we already have plevel to specify overriding rule.

Previously we check the plevel, if the plevel have higher priority level then we can override the attribute

zxybazh replied:
Yeah that make sense. Thanks for the explanation.

tqchen commented on src/target/llvm/intrin_rule_hexagon.cc:
```diff
@@ -19,44 +19,54 @@
-TVM_REGISTER_GLOBAL("tvm.intrin.rule.hexagon.exp")
-    .set_body(DispatchLLVMPureIntrin<::llvm::Intrinsic::exp, 1>);
+TVM_REGISTER_OP("tir.exp").set_attr<FLowerIntrinsic>(
+    "hexagon.FLowerIntrinsic", PackedFunc(DispatchLLVMPureIntrin<::llvm::Intrinsic::exp, 1>));
```
would be nice if we can modify the signature of DispatchLLVMPureIntrin to be PrimExpr->PrimExpr, so we do not have to do the convertion to PackedFunc

zxybazh replied:
Thanks! Fixed in the new commit.

tqchen commented on src/target/llvm/intrin_rule_llvm.cc:
```diff
@@ -25,155 +25,175 @@
+                               PackedFunc([](const TVMArgs& targs, TVMRetValue* rv) {
+                                 using tir::make_const;
+                                 using tir::make_zero;
+                                 PrimExpr e = targs[0];
```
Would be great use the typed version of PackedFunc

zxybazh replied:
It seems that in order to use typed version of `PackedFunc` the signature of the later function must be exactly the same, i.e., changing signatures of all functions to `PrimExpr(PrimExpr)`. I think the current functions are pretty much in good shape and if we want to change the function type of `FLowerIntrinsic` in the future we don't have to modify the code. What do you think?

tqchen replied:
The original API was built when we did not have `PrimExpr(PrimExpr)`. Given that I don;t think the FLowerIntrinsic signature will change in the future, it would be great to switch to the typed version

zxybazh replied:
I see. In that case typed version would be better. Thanks for pointing that out.

tqchen reviewed (changes requested):
two minor nits mainly on the api and use TypedPackedFunc over PackedFunc. Overall LGTM!

tqchen reviewed (changes requested):
Thanks for keep improving the PR! Taking a close look this time. I think we can merge once the current set of comments are addressed

zxybazh commented on src/tir/transforms/lower_intrin.cc:
```diff
@@ -42,28 +42,38 @@ class IntrinInjecter : public tvm::arith::IRMutatorWithAnalyzer {
-      if (r.defined()) return r;
+      for (const std::string& pattern : patterns_)
+        if (Op::HasAttrMap(pattern)) {
+          auto f_lower_intrin_map = Op::GetAttrMap<FLowerIntrinsic>(pattern);
```
Good point, fixed in the new commit.

tqchen reviewed (approved):
LTGM! Thanks @zxybazh !

junrushao1994 commented:
Thanks @zxybazh @tqchen!
